src/world/World.php

```diff
@@ -588,6 +599,135 @@ private function initRandomTickBlocksFromConfig(ServerConfigGroup $cfg) : void{
 		}
 	}
 
+	public function getWeather() : WeatherType{
+		return $this->weather;
+	}
+
+	public function setWeather(WeatherType $type, int $duration = 6000) : void{
+		if($this->weather === $type){
+			$this->weatherDuration = $duration;
+			return;
+		}
+
+		$ev = new WeatherChangeEvent($this, $this->weather, $type);
+		$ev->call();
+		if($ev->isCancelled()){
+			return;
+		}
+
+		$this->weather = $type;
+		$this->weatherDuration = $duration;
+		$this->weatherTick = 0;
+
+		$this->applyWeather();
+		$this->broadcastWeatherPackets();
+	}
+
+	private function tickWeather() : void{
+		$this->rainLevel = max(0.0, min(1.0, $this->rainLevel + (mt_rand(-2, 2) / 100)));
+		$this->thunderLevel = max(0.0, min(1.0, $this->thunderLevel + (mt_rand(-1, 1) / 100)));
+		$this->weatherTick++;
+
+		if($this->weatherTick >= $this->weatherDuration && $this->weatherDuration > 0){
+			$this->autoChangeWeather();
+		}
+
+		if($this->weather === WeatherType::THUNDER && mt_rand(0, 200) === 0){
+			$players = $this->getPlayers();
+			if(count($players) > 0){
+				$target = $players[array_rand($players)];
+				$spawn = $target->getPosition();
+				$x = (int) $spawn->x + mt_rand(-30, 30);
+				$z = (int) $spawn->z + mt_rand(-30, 30);
+				$y = $this->getHighestBlockAt($x, $z) ?? max(64, (int) $spawn->y);
+
+				$this->spawnLightning(new Vector3($x, $y, $z));
+			}
+		}
+	}
+
+	private function autoChangeWeather() : void{
+		$roll = mt_rand(1, 100);
+		if($roll <= 70){
+			$this->setWeather(WeatherType::CLEAR, mt_rand(6000, 12000));
+		}elseif($roll <= 90){
+			$this->setWeather(WeatherType::RAIN, mt_rand(4000, 9000));
+		}else{
+			$this->setWeather(WeatherType::THUNDER, mt_rand(2000, 6000));
+		}
+	}
+
+	private function applyWeather() : void{
+		match($this->weather){
+			WeatherType::CLEAR => (function(){
+				  $this->setRainLevel(0.0);
+				  $this->setThunderLevel(0.0);
+			})(),
+			WeatherType::RAIN => (function(){
+				$this->setRainLevel(1.0);
+				$this->setThunderLevel(0.0);
+			})(),
+			WeatherType::THUNDER => (function(){
+				$this->setRainLevel(1.0);
+				$this->setThunderLevel(1.0);
+			})(),
+		};
```

Comment:
As I said before, a switch would be better for this, instead of baking closures inline